### ISSUE TITLE
OSX Finder hang on copy and server DoS

### ISSUE THREAD

**Issue by jonasbardino** [mar10/wsgidav#15]
We're working on replacing pywebdav with wsgidav as the provider of our WebDAV(S) interface to user files on the Minimum intrusion Grid (www.migrid.org). It looks very promising so far, with much better performance, protocol compliance and native Windows 7+ support.
One draw-back, however, is an issue with native Mac OSX support. The OSX file manager, Finder, connects and operates alright, but when copying a file from one DAV directory to a sub-directory, it hangs and leaves the server spinning at 100% CPU - effectively slowing it to a near DoS state.

We run our own instance behind SSL with some customization, but we also tested with the wsgidav server in a very basic configuration and the problem remains:
wsgidav --host=0.0.0.0 --port=8021 --root=./davshare --debug > debug-copy-hang.log 2>&1
Apparently I can't attach the resulting log here but it's available at:
http://www.migrid.org/debug-copy-hang.log
Similar log for the same (working) operation from windows 8 file manager and Linux with Thunar/gvfs are available at:
http://www.migrid.org/debug-copy-win8-okay.log
http://www.migrid.org/debug-copy-gvfs-okay.log

I'm not a DAV expert, but it looks strange to me that Finder locks both the qwe.txt and ._qwe.txt and then only unlocks the latter (near the end of the log). It might be a Finder bug, but it should probably not leave the server spinning then?
Windows locks and unlocks only once and gvfs does not lock at all, so it makes sense that it works without problems there.

I don't have a Mac myself, but I can ask a colleague to help with further testing, if you can suggest additional things to try in order to track down and solve the problem?

**Comment by jonasbardino** [comment]
Additional notes: 
For whatever reason issuing the similar 'cp' on the command-line inside the mount succeeds. The corresponding log indicates that only the qwe.txt file is locked and unlocked in that case, not the ._qwe.txt meta data file:
http://www.migrid.org/debug-cp-osx-okay-maxlocklog.log

The issue is not specific to internal copy, btw, it's the same result when uploading a local file using Finder.

**Comment by mar10** [comment]
Great to hear you are planning to use it.
I assume that you access the server from another machine, so the 100% CPU usage is definitely caused by WsgiDAV (not Finder)...
Could be a recursion / endless loop in the LockManager.
Can you set breakpoints there or add some logging?

I will try to look into it, if I find some time, but I cannot promise that this will be very soon...

**Comment by jonasbardino** [comment]
Thanks for the quick response!
Yes indeed, the server runs CentOS 6.6 and the remote client OSX. It is 'top' showing the wsgidav process running at 100% cputime and other clients get severely slowed then.
I also tried with locksmanager=False for the daemon to see if that would work around the problem, but then the share just effectively becomes read-only.
I'm not familiar with the LockManager code at all, but I can take a look tomorrow. Any further tests may have to wait until next week, though.

**Comment by jonasbardino** [comment]
I've borrowed a Mac (OSX 13.4) and with extra logging in the lock_manager and request_server I've tracked the busy wait down to be not in the locking but rather in the chunked upload in doPUT. More specifically the server stalls when calling the wsgi input readline in line 670 of request_manager.py:
    buf = environ["wsgi.input"].readline()

so the log call I added right after that line never fires. 

AFAICT the readline should receive a line with the next chunk size, but it just hangs there with full cpu use:
http://www.migrid.org/debug-copy-osx-13.4.0-hang-maxlock-putlogger.log
(last line containing doPUT)

The Finder dialog stalls for a while saying zero of 4 bytes uploaded:
http://www.migrid.org/osx-13.4.0-copy-stall.png
before eventually timing out with the error:
http://www.migrid.org/osx-13.4.0-copy-hang.png

The HTTP headers do not include Content-length for that request, so I guess the client *should* send something. It does have a HTTP_X_EXPECTED_ENTITY_LENGTH: '4' header, though. So maybe it tries to tell us the chunk size that way.
Any ideas about what to try?

---

As a side note, these two sound related, but with no solution so far:
https://groups.google.com/forum/#!topic/seafile/mTyD1tBGw_o
https://groups.google.com/forum/#!topic/seafile/Mvu8p4xmokQ
I suppose Seafile webdav relies on wsgidav based on issue #14 .

**Comment by jonasbardino** [comment]
It actually looks like just using the HTTP_X_EXPECTED_ENTITY_LENGTH instead of reading the chunk length with the readline works for my example!
I've tested some more and it appears to solve the copy and upload issues for all files I've thrown after it (up to 256MB so far).
So maybe you can comment on the idea of changing the readline to something like:
                if "Darwin" in environ.get("HTTP_USER_AGENT", "") and \
                        environ.get("HTTP_X_EXPECTED_ENTITY_LENGTH", ""):
                    buf = environ.get("HTTP_X_EXPECTED_ENTITY_LENGTH", "")
                else:
                    buf = environ["wsgi.input"].readline()

Not sure if it should be OSX/Darwin specific, but I haven't seen that behaviour with any other clients.
I'll be mostly offline until next week, but will promise then if needed.

**Comment by mar10** [comment]
Very cool, thanks for tracking that down.
I was able to reproduce it yesterday, but haven't haven't had time to investigate furrther. So your input is very helpful.

Maybe we can avoid user agent sniffing and simply use `HTTP_X_EXPECTED_ENTITY_LENGTH` if present, and no `CONTENT_LENGTH` was passed.
I will try a bit over the weekend...

**Comment by jonasbardino** [comment]
Sounds good!
Sure, if you are confident it won't cause regressions for non-OSX clients, there's no need to include the agent check.
One curious thing I noted is that  it is only the first readline call there that hangs. The second one a bit further down in the code works fine when left as is. Maybe because the client never enters the multiple chunks in the same connection from OSX, so that it always receives an EOF or something there.
I'll start a gigabyte+ upload now to make sure big files behave, too. Just ping me if you want me to test anything else.

**Comment by mar10** [comment]
Could you check if the big file identical after copy, e.g. by calculating a checksum?

I also found this reference https://github.com/fruux/sabre-dav/blob/master/lib/DAV/CorePlugin.php#L452
Seems that there may be a theoretical edge case, when the file is modified/appended while copying.

This is how i applied your  patch. It works for me in a quick test, but I guess you tested more thoroughly:
```py
if environ.get("HTTP_TRANSFER_ENCODING", "").lower() == "chunked":
    # Chunked Transfer Coding
    # http://www.servlets.com/rfcs/rfc2616-sec3.html#sec3.6.1

    if environ.get("HTTP_X_EXPECTED_ENTITY_LENGTH"):
        # Probably Mac Finder, that does not prepend chunk-size + CRLF
        # (which is not compliant with the spec)?
        buf = environ.get("HTTP_X_EXPECTED_ENTITY_LENGTH", "")
    else:
        buf = environ["wsgi.input"].readline()
```

**Comment by jonasbardino** [comment]
The biggest file I tried so far was just a load of zeros, but md5sum matched on the server after upload. 

The patch looks good to me. I manually applied it and tested again now uploading a 1.1GB Linux iso and a smaller file - both resulted in md5sum match, too.

That corner case sounds rather nasty. I'm not sure how easy it is to test in practice, but I'm mostly worried that it could leave the 2nd readline spinning if Finder ends up sending more than HTTP_X_EXPECTED_ENTITY_LENGTH bytes. 
I tried uploading a 93MB file and appended the same file contents to it during upload. It still finished successfully with just the original version uploaded. So that didn't seem to trigger any changes. 
Anything else I should try?

**Comment by mar10** [comment]
I only assumed this corner case from the SabreDav comment, not sure if it is relevant at all.
If you are curious, you might decrement the HTTP_X_EXPECTED_ENTITY_LENGTH value, to simulate that behavior.

If I understood correctly, omitting the chunk-size is a violation of rfc2616, so maybe we should add the agent check for 'Darwin' that you originally proposed.

Let me know if you want to create a Pull Request, which would add you to the contributors hall of fame ;-)
And thanks for contributing!

**Comment by jonasbardino** [comment]
Cool, I've made a pull request now (my first ever, so I hope I haven't messed up anything :-) )
I added the Darwin check again as you suggested.

**Comment by mar10** [comment]
Thanks for contributing!

**Comment by jonasbardino** [comment]
You're welcome, and thanks for your speedy feedback!

I played a bit more around with decrementing/incrementing the actual payload size as you suggested and found a couple of problems with the patch, though:
 * HTTP_X_EXPECTED_ENTITY_LENGTH is the *integer* size so using it directly as buf value and thus later hex-to-int converting actually makes the length value too big. Only by coincidence that's not a problem because the read call stops on EOF anyway. It is also easy to fix by moving the conversion inside the else case.
 *  Even then, when I manually decrement the resulting l-value below payload size, the next readline does indeed end up in the spin I originally worried about.

I'm working on a better fix to make the l-loop run until the advertised number of bytes were read or read returned the empty string due to EOF. The latter will again be OSX specific to keep rfc-compliance for other clients. 

I need some more testing, but I will make a new pull request later.

**Comment by ysoftman** [comment]
Hi, I have the same problem below code when I request 'PUT' to wsgidav server by aiohttp client.
```python
buf = environ["wsgi.input"].readline()
```

[my test code](https://github.com/ysoftman/test_code/blob/4be0215d5d2a896894626ce470f3c96ac35bf770/python/progress_bar_tqdm.py#L111)
In my test code using aiohttp there is no  X_EXPECTED_ENTITY_LENGTH header and User-Agent was 'Python/3.9 aiohttp/3.7.4.post0'

So I set headers X_EXPECTED_ENTITY_LENGTH: size_of_chunk, User-Agent: Darwin, and It works.
Because I understand omitting the chunk-size is a violation of rfc2616, I used the HTTP_X_EXPECTED_ENTITY_LENGTH header.
But why does the code check user-agent: Darwin only in 
https://github.com/mar10/wsgidav/blob/de622d0eaef96c63109af8501caf4c8c25f0f0a4/wsgidav/request_server.py#L670-L672
